### PR TITLE
Add descriptive docstring to TemperatureLogitsWarper

### PR DESCRIPTION
**Pull request by nablabits**
# What does this PR do?

Fixes one of the cases in https://github.com/huggingface/transformers/issues/24783

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

## Who can review?
@gante

## Notes
My first PR to this library, greatly appreciate your patience :wink:

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -172,11 +172,73 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
     r"""
-    [`LogitsWarper`] for temperature (exponential scaling output probability distribution).
+    [`LogitsWarper`] for temperature (exponential scaling output probability distribution) which effectively means that
+    we can control the randomness of the predicted tokens.
+
+    <Tip>
+
+    Make sure that `do_sample=True` is included in the `generate` arguments otherwise the temperature value won't have
+    any effect. As noted in `generation_strategies#Assisted-Decoding`, this parameter along with assisted decoding will
+    help improving the latency.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:182]:
```suggestion
    any effect. 
```

Assisted decoding is still niche, so let's not add this detail (yet) 😉

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -172,11 +172,73 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    </Tip>
 
     Args:
         temperature (`float`):
-            The value used to module the logits distribution.
+            The value used to module the logits distribution. A value of `1.` will make no difference in the output
+            with respect to calling `generate` without temperature.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:189]:
```suggestion
            The value used to module the logits distribution. A value smaller than `1` decreases randomness (and vice versa), with `0` being equivalent to shifting all probability mass to the most likely token.
```

Let's focus on the relevant use cases (temperature < 1) rather than on the caveats :)

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -172,11 +172,73 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    Examples:
+
+    ```python
+    >>> import torch
+    >>> from transformers import AutoTokenizer, AutoModelForCausalLM
```
Comment (gante) [inline src/transformers/generation/logits_process.py:195]: Uhmm... I have 2 requests for the example:
1. Let's avoid nested functions, to make the example as plain as possible
2. Let's try to craft an example with 20-30 lines

Let your creative juices go all out here :) A viable alternative could be e.g. to show what happens when you call generate with a different seed, and that a low temperature leads to seed-invariant results

**Review by gante — COMMENTED** [review]
Thank you for opening the PR 🙏 

A few suggestions to make these docs more consistent with the other docs

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (nablabits) [reply src/transformers/generation/logits_process.py]: Cool, thanks for your time :+1:

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (nablabits) [reply src/transformers/generation/logits_process.py:193]: Ahh, yeah, I had a kind of voice inside me whispering: *don't put that super-duper function there...*

Let me see what can take out of this poor brain :brain: I have :wink:

Comment (nablabits) [reply src/transformers/generation/logits_process.py:195]: Finally I couldn't squeeze my neurons to get something beyond your suggestions, hope the pr looks ok now, but don't loose the hope, eventually  we will get there :wink:

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -172,11 +172,58 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class TemperatureLogitsWarper(LogitsWarper):
     r"""
-    [`LogitsWarper`] for temperature (exponential scaling output probability distribution).
+    [`LogitsWarper`] for temperature (exponential scaling output probability distribution) which effectively means that
```
Comment (gante) [inline src/transformers/generation/logits_process.py:175]:
```suggestion
    [`LogitsWarper`] for temperature (exponential scaling output probability distribution), which effectively means that
```
nit :)

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -172,11 +172,58 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    Make sure that `do_sample=True` is included in the `generate` arguments otherwise the temperature value won't have
+    any effect.
+
+    </Tip>
 
     Args:
         temperature (`float`):
-            The value used to module the logits distribution.
+            The value used to module the logits distribution. A value smaller than `1` decreases randomness (and vice
```
Comment (gante) [inline src/transformers/generation/logits_process.py:187]:
```suggestion
            Strictly positive float value used to modulate the logits distribution. A value smaller than `1` decreases randomness (and vice
```

We can remove the last sentence (see below) and make it more compact this way :) Also note that `module` should be `modulate`

`make fixup` may need to be executed after this change!

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -172,11 +172,58 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+            versa), with `0` being equivalent to shifting all probability mass to the most likely token. However, bear
+            in mind that a positive float should be given, therefore, `float(0)` is not allowed.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:189]:
```suggestion
            versa), with `0` being equivalent to shifting all probability mass to the most likely token. 
```

**Review by gante — APPROVED** [review]
That is a great example @nablabits! 🔥 

Thank you for iterating on it, I've requested a few minor changes (to further improve information density), and it should be ready to merge after they are addressed 🤗

**Comment by nablabits** [comment]
> That is a great example @nablabits! :fire: 
> 
> Thank you for iterating on it, I've requested a few minor changes (to further improve information density), and it should be ready to merge after they are addressed :hugs: 

Hi @gante, thanks for your patience, guidance and support, much appreciated :hugs: . I greatly enjoyed the learning experience. Are you happy for me to pick something else in the list (I'd like to deepen in my knowledge of this tiny bit of the platform) or the protocol suggests that I should leave remaining tasks for other folks?

**Comment by gante** [comment]
@nablabits feel free to pick more tasks from the list, as many as you want (one at a time, of course) -- as long as you confirm that no one is working on a given task and that you share on the issue that you've decided to take it 🤗

**Review by sgugger — APPROVED** [review]
Thanks!
